rclc/QUALITY_DECLARATION.md

```diff
@@ -0,0 +1,200 @@
+This document is a declaration of software quality for the `rclc` package, based on the guidelines in [REP-2004](https://www.ros.org/reps/rep-2004.html).
+
+# `rclc` Quality Declaration
+
+The package `rclc` claims to be in the **Quality Level 2** category when it is used with a **Quality Level 2** middleware.
+
+Below are the rationales, notes, and caveats for this claim, organized by each requirement listed in the [Package Quality Categories in REP-2004](https://www.ros.org/reps/rep-2004.html).
+
+## Version Policy [1]
+
+### Version Scheme [1.i]
+
+`rclc` uses `semver` according to the recommendation for ROS Core packages in the [ROS 2 Developer Guide](https://docs.ros.org/en/rolling/Contributing/Developer-Guide.html#versioning).
+
+### Version Stability [1.ii]
+
+`rclc` is at a stable version, i.e. `>= 1.0.0`.
+The current version can be found in its [package.xml](package.xml), and its change history can be found in its [CHANGELOG](CHANGELOG.rst).
+
+### Public API Declaration [1.iii]
+
+All symbols in the installed headers are considered part of the public API.
+
+All installed headers are in the [`include`](./include/rclc) directory of the package, headers in any other folders are not installed and considered private.
+
+### API Stability Policy [1.iv]
+
+`rclc` will not break public API within a released ROS distribution, i.e. no major releases once the ROS distribution is released.
+
+### ABI Stability Policy [1.v]
+
+`rclc` contains C code and therefore must be concerned with ABI stability, and will maintain ABI stability within a ROS distribution.
+
+### ABI and ABI Stability Within a Released ROS Distribution [1.vi]
+
+`rclc` will not break API nor ABI within a released ROS distribution, i.e. no major releases once the ROS distribution is released. Any ABI break in **rclc** will be done between minor versions and it should be clearly stated in the release notes.
```

Comment:
*"Any ABI break in **rclc** will be done between minor versions and it should be clearly stated in the release notes."*
Not sure I understand what this means. Does this mean minor version bump for every ABI break? How does this relate to the statement above, *"maintain ABI stability within a ROS distribution"*.
Maybe I'm confused and don't read it the right way. :D

Reply:
*"Any ABI break in **rclc** will be done between minor versions and it should be clearly stated in the release notes."
Not sure I understand what this means. Does this mean minor version bump for every ABI break?* 

Yes. it means any API change requires bumping major version. Any ABI change requires bumping minor version.

*How does this relate to the statement above, *"maintain ABI stability within a ROS distribution"*.
Maybe I'm confused and don't read it the right way. :D*

I am also not sure, what is meant with "maintain ABI stability" if changing minor version is allowed. The same sentence is in [rcl/QUALITY_DECLARATION.md](https://github.com/ros2/rcl/blob/master/rcl/QUALITY_DECLARATION.md) and [rclcpp/QUALITY_DECLARATION.md](https://github.com/ros2/rclcpp/blob/master/rclcpp/QUALITY_DECLARATION.md)

I will moved the sentence to the subsection of ABI stability. There is even still the type "ABI and ABI Stability" in rcl and rclcpp QUALITY_DECLARATION instead of "API and ABI Stability".

Reply:
But to be honest, bug-fixing and updating the functionalty will be inevitable within a released ROS distribution. This means the minor version will be changed. 

Therefore, in my view, the sentence should changed to: 

`rclc` will not break API within a released ROS distribution, i.e. no major releases once the ROS distribution is released. ABI can be changed with minor versions. 

I don't know what the guys from OSRF mean by "rclcpp will not break API nor ABI within a released ROS distribution". because any bug-fix will break ABI.

@norro @ralph-lange Do you agree with the change?

Reply:
Update: 

I read this [ROS Developer Guide](https://docs.ros.org/en/foxy/Contributing/Developer-Guide.html). 
From this I conclude, that neither API breaks nor ABI breaks are allowed within a ROS distribution. bumping minor-version within a release is allowed, if it does not break the ABI.

ABI is the compiler-interface, e.g. adding new fields, removing functions etc is not allowed. There is also this  [tool](https://lvc.github.io/abi-compliance-checker/) to check ABI stability.

Reply:
This makes more sense to me. Than the sentence above is really at least misleading, if not wrong (also in rcl and rclcpp).

Reply:
Updated the description. In my view this issue is resolved.